src/events/processCollectedSignatures/estimateGas.js

```diff
@@ -0,0 +1,51 @@
+const { HttpListProviderError } = require('http-list-provider')
+const {
+  AlreadyProcessedError,
+  IncompatibleContractError,
+  InvalidValidatorError
+} = require('../../utils/errors')
+const { parseMessage } = require('../../utils/message')
+
+async function estimateGas({
+  foreignBridge,
+  validatorContract,
+  message,
+  numberOfCollectedSignatures,
+  v,
+  r,
+  s,
+  address
+}) {
+  try {
+    const gasEstimate = await foreignBridge.methods
+      .executeSignatures(v, r, s, message)
+      .estimateGas()
+    return gasEstimate
+  } catch (e) {
+    if (e instanceof HttpListProviderError) {
+      throw e
+    }
+
+    const requiredSignatures = await foreignBridge.methods.requiredSignatures().call()
+    if (requiredSignatures.toString() !== numberOfCollectedSignatures.toString()) {
+      throw new IncompatibleContractError('The number of collected signatures does not match')
+    }
+
```

Comment:
If you look at the contract code you will see that every signature in the message is checked to be sure that:
* the corresponding signature was made by a validator;
* this signature is mentioned only once in the array;

```solidity
        for (uint256 i = 0; i < requiredSignatures; i++) {
            address recoveredAddress = ecrecover(hash, _vs[i], _rs[i], _ss[i]);
            require(_validatorContract.isValidator(recoveredAddress));
            if (addressArrayContains(encounteredAddresses, recoveredAddress)) {
                revert();
            }
            encounteredAddresses[i] = recoveredAddress;
        }
```
So let's add this check as well. The check that we will get a singature which belongs to an account and it is not validator any more is still probable. You could even consider to add this check just after the line:

```javascript
const signature = await homeBridge.methods.signature(messageHash, index).call()
```

Reply:
In this case, maybe is better to just ask the contract. The problem is that this logic is in the `Message` library and we can't call it directly. The same thing happened to me with the `parseMessage` logic.

Since it's the third time that this happens, would it make sense to expose all the library functions somehow? I see two ways in which this could be done:

1. Make all methods `internal` and add methods with the same name in the bridge contracts that just call the library.
2. Deploy and link the library and drop the `internal`s.

**2** looks better to me, but it would mean modifying the deploy script.

Reply:
I would avoid big amount of changes on the contracts side since we already had a security audit for them. 

If we just ask the contract we still get no point why we fails since two check are combined there:
```solidity
require(_validatorContract.isValidator(recoveredAddress));
```
 and
```solidity
if (addressArrayContains(encounteredAddresses, recoveredAddress)) {
    revert();
}
```

My suggestion is the following:
* Since we generate the list of signatures before passing it to the contract, it is not necessary to check for signature duplication in `processCollectedSignatures/estimateGas.js`.
* But we need to make sure that the signature in the list is for an actual validator. That's why we need to restore every validator address from the signatures we received and call `isValidator` for them.

Reply:
Yes, at least in the nearest future (release of `native-to-erc`, `erc-to-erc` and `erc-to-native`) we will not change contracts a lot.

Reply:
I passed the signatures to the `estimateGas` function to keep the structure of error checking. There it recovers the address for every signature and checks that is from a validator.